akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala

```diff
@@ -170,16 +172,24 @@ import scala.compat.java8.FutureConverters._
     }
 
     (stageLogic, new SourceQueueWithComplete[T] {
+      @volatile
+      var completed = false
       override def watchCompletion() = completion.future
       override def offer(element: T): Future[QueueOfferResult] = {
-        val p = Promise[QueueOfferResult]
-        stageLogic.invoke(Offer(element, p))
-        p.future
+        if (!completed) {
+          val p = Promise[QueueOfferResult]
+          stageLogic.invoke(Offer(element, p))
+          p.future
+        } else {
+          Future.successful(QueueOfferResult.QueueClosed)
+        }
       }
       override def complete(): Unit = {
+        completed = true
         stageLogic.invoke(Completion)
       }
       override def fail(ex: Throwable): Unit = {
+        completed = true
```

Comment:
should this be handled differently to actually return `QueueOfferResult.Failure` on call to `offer`?